### ISSUE TITLE
Rate limit PR creation rate

### ISSUE THREAD

**Issue by rarkins** [renovateapp/renovate#1363]
Although the onboarding PR + subsequent Pin Dependencies PR are nice and orderly, the post-pinning period can be a bit overwhelming. Although the onboarding PR gives you plenty of warning, it can be overwhelmimg to have 10, 20 or 30+ PRs raised all at once. This can also consume CI resources non-stop, particularly if the repo applies a policy of rebasing PRs that are not up-to-date with `master`.

Although it's inevitable that the project needs to get those PRs one way or another, it is desirable if we can:
- Reduce the amount of "retesting" caused by rebasing
- Reduce the concurrently tested PRs so as not to clog up the CI queue for non-Renovate PRs

A suggestion by @alexeagle is to *rate limit* PRs. Doing so at least gives the opportunity for a repository maintainer to receive, test and merge PRs before the next ones arrive. If done slow enough, you might even end up not needing too many rebases.

Still to decide:

- [ ] Should this be exclusively during onboarding, or as an ongoing setting?
  * the idea is to make it configurable, so if we make it ongoing then they at least can remove or adjust it once over the onboarding period

- [ ] Is an integer configuration for per hour enough granularity? 
  * I usually think in terms of running once per hour, and our scheduling also works per-hour
  * This means slowest figure would be one per hour

- [ ] What default value?
  * e.g. default to unlimited - like today - or default to a low amount like 2 per hour?

- [ ] Should Renovate "break" the PR rate if existing ones have been merged?
  * e.g. even if you configure 1 per hour, if you merge the first one after 35 minutes then wouldn't it be desirable to get a new one created immediately after, if possible?

**Comment by alexeagle** [comment]
My opinions:
- keep the setting all the time just so my mental model of how renovate works doesn't change after onboarding
- make it an integer of PRs per day so that it's easy to express fractions of an hour in case someone wants that, I'd probably pick 24
- I don't have a strong feeling about default, I suppose for a small repo you just want to configure renovate and be done with it, so novice user probably prefers the current behavior
- I think it's easier to understand if the rate limiting doesn't interact with any user behavior

**Comment by rarkins** [comment]
I wonder if your "per day" suggestion would be too confusing. You're suggesting "x PRs per day" but actually intending it as "x/24 PRs per hour", right?

If we want fractions in an hour them perhaps we can make it minutes separation between PRs, e.g. you would pick 60.

**Comment by edmorley** [comment]
Another advantage of this feature: It gives the user a chance to fix the Renovate config if it turns out the commit message/... style aren't to their liking after all.

> Should this be exclusively during onboarding, or as an ongoing setting?

I think the feature is less important post-onboarding, since by then (a) people watching the repo are accustomed to Renovate, so less likely annoyed by the PRs, (b) there will be fewer PRs going forwards so less likely the threshold would be reached anyway (at least for projects that don't severely restrict Renovate to say only once a month), (c) there's less chance of the user not being happy with their Renovate commit message style etc.

Also worth considering that for some people, trickling PRs gradually post-onboarding (when they are no longer actively looking at the PRs / in the "let's update things" mindset) actually might be "more" noisy, since they get email notifications spread out over a longer period of time.

> Is an integer configuration for per hour enough granularity?

I think so, yeah.

> What default value?

So long as the messaging to users is clear (to avoid "why hasn't Renovate opened PRs for things that are out of date yet, is it broken?"), then I'd say it should default to throttled, though perhaps not too aggressively. Considerations:
* not all of the PRs may pass tests, so it might be annoying if so few PRs were opened initially that the user had nothing they could actually merge straight away
* perhaps there is some way to reduce merge conflicts? (eg don't open PRs in alphabetical order of dependencies, since they'll be next to each other in `package.json` and `yarn.lock`, so likely to conflict)
* should `dependencies` or `devDependencies` be updated first? Perhaps the latter, as the user can more easily merge `devDependencies` since less risky (or in the case of say eslint, tests are sufficient), so (a) it gives the user some satisfaction at having merged something, (b) they'll be merged and out the way by the time other PRs come along that might have conflicted and meant a rebase required.

**Comment by edmorley** [comment]
Ah sorry forgot to add: For a rate limit of say 4 an hour, should that be 4 opened at the same time at the start of the hour, or 1 opened every 15 minutes?

**Comment by destroyerofbuilds** [comment]
> Another advantage of this feature: It gives the user a chance to fix the Renovate config if it turns out the commit message/... style aren't to their liking after all.

👍 

> e.g. even if you configure 1 per hour, if you merge the first one after 35 minutes then wouldn't it be desirable to get a new one created immediately after, if possible?

For on-boarding only PRs, I would prefer the approach that, upon accepting a pinned pull request, I would receive the next pull request. My reasoning is that it means I remain in the PR interface accepting pinned pull requests, instead of context switching to another task for some amount of time, only to then break away, context switch, back, again, and again.

> So long as the messaging to users is clear (to avoid "why hasn't Renovate opened PRs for things that are out of date yet, is it broken?")

^ definitely. Running an on-premise instance of Renovate, this question comes up quite often (And we run Renovate every 10 minutes)

**Comment by rarkins** [comment]
Good suggestions and questions.

As I mentioned earlier, Renovate's existing `schedule` capability restricts to maximum "hourly" start/stop. That way such schedules can be easily fulfilled so long as Renovate runs at least once per hour per repository.

For the app, the way it works is:

- There is a pool of "workers" who pluck jobs off a queue non stop
- Once per hour, all repositories are added to a best-effort queue if they're not already in it
- Select types of webhooks or manual invocations from me result in a repository being added or moved in the queue with a higher priority, so it gets serviced by a worker usually immediately

So this type of approach is not easily compatible with something like "every 15 minutes" (i.e. 4 per hour spread evenly).

Also, because of the above characteristics (Renovate may run multiple times in an hour) we also can't use simple logic like "if they configure 2 per hour then create at most 2 every run".

Which means that every time Renovate runs, it needs to work out "how many are appropriate to create right now?"

**Comment by rarkins** [comment]
I've decided that the best way to achieve this is to simply use hourly windows, e.g. :00-:59 and rate limit within those. That way Renovate can just use the logic:

- Is there a PR limit in place?
- What hour is this?
- How many PRs already created this hour?

**Comment by rarkins** [comment]
I have implemented this feature as described above. i.e. the rate limiting is per "calendar hour" rather than "last 60 minutes", to avoid race conditions.

The config is `prHourlyLimit` and is of type integer, meaning that 1 per hour is the slowest you can configure. The app itself defaults to 0 (unlimited) but I decided to update my "config:base" preset to default to 2 per hour, that way all new onboarding repos will get 2 PRs per hour maximum without needing any further configuration. Anyone who wants faster or slower than that can update the `renovate.json` configuration in the onboarding PR to get the result they want.

**Comment by edmorley** [comment]
Sounds great - many thanks!
